### DISCUSSION TITLE
Computationally heavy calculation

### DISCUSSION THREAD

**Discussion by cesarecaoduro** [donalffons/opencascade.js#280] [Q&A]
Good morning everyone,
I hope someone can help me with this topic, before I move forward with different solutions.
I have been trying to build a configurator using opencascadejs, and I have hit a roadblock now.

One of the operation is to calculate a huge amount of points along a wire at certain distances.
```
function pointAtLengths(wire: TopoDS_Wire, ts: number[]): Point3[] {
        var points: Point3[] = []
        var adaptor = new _occt.BRepAdaptor_CompCurve_2(wire, true)
        ts.forEach(t => {
            var abscissaPoint = new _occt.GCPnts_AbscissaPoint_3(1e-6, adaptor, t, 0.0)
            var pt: gp_Pnt = new _occt.gp_Pnt_1()
            adaptor.D0(abscissaPoint.Parameter(), pt)
            points.push(fromGpPnt(pt))
        })
        return points
    }
```
The version of this method that runs in c# is basically instantaneous, but the one that runs in the browser, basically locks completely the app. Now I know that I can build a worker to avoid blocking the main thread, but this is not the point.

The wire is made of a series of edges (mostly lines and arcs), and could be long a few thousand of meters. I am basically trying to calculate a point every meter along that wire which represents a road alignment.

![image](https://github.com/user-attachments/assets/992aeef2-65ed-481a-a0d0-6a7a4955461e)

Is there anything I am doing wrong?

**Comment by cesarecaoduro** [comment]
@donalffons more than happy to make a donation if I can get some help in here.
My app works in ```Nuxt3```

**Reply by matas-bitbybit-dev** [reply]
Hey @cesarecaoduro I'd love to help, but I'm not familiar with Nuxt3. If I watch your error I'd guess it's not really related to the point subdivision operation, if I understand it your app fails to initialize due to memory limit.

Does your app succeed initializing without running pointAtLengths function?

I'm not so sure what can be the cause of this, but if you run occt on Node you could try increasing allocated memory to the process.

https://stackoverflow.com/questions/34356012/how-to-increase-nodejs-default-memory

If it is in fact caused by pointAtLengths function I could run some tests on https://bitbybit.dev if you would provide a step file of your  wire and see wether I would run into something similar

**Comment by cesarecaoduro** [comment]
I don’t believe the issue is Nuxt. Nuxt is only a framework on top of Vue.
It seems like that as soon as I try to calculate many points along a wire is stops working.
The wire is not that complex. I can send you a brep tomorrow, but I think any curve should do the trick. Maybe you can try a sin curv lo g 3-4000 meters and then try to calculate a point along that curve every meter.

I’ve also tried to use your implementation but I’m running into some issue with Vue. Have you tried already? An example of bitbybit with Vue would be unbelievably helpful :)

You’ve done an amazing job with the platform.

**Reply by matas-bitbybit-dev** [reply]
If I understand it well, Nuxt offers server side rendering (SSR) capabilities to Vue and thus runs a NodeJS server. Running OCCT in NodeJS differs quite a bit from running it in the browser. I could provide pure Vue example, but running occt in SSR apps is beyond me at this point ;)

I just did a quick test, probably not exactly as you defined, but it's a wire that is around 4000 units long made out of linear and arc edges, which is subdivided to 4000 points...

https://bitbybit.dev/app/matas/iQv4qa5sTH8oVXiMKI0X/OboM0oGTHoQf27LDfeWS?editor=rete

But I would really advise to look into the error itself, it says the error occured while initializing the application and complains about the memory allocation.

I'm gonna be off for the rest of the weekend :)

**Reply by cesarecaoduro** [reply]
Thanks…I’ll run some more tests.
Btw I’ve disabled SSR so it should run all on the client side.
I’ll try with a normal typescript project scaffolded with Vite.

**Reply by cesarecaoduro** [reply]
It looks like I can get it to work (marked the component as client only), the problem is still performances.
When I run it using pythonocc or a c# binding of opencascade, the 2000 points are calculated instantaneously, while with JavaScript it hangs for a good amount of time.

**Reply by cesarecaoduro** [reply]
I have uploaded a repo with the test, using your library. I think I have hooked that up, but I don't see anything happening.
https://github.com/cesarecaoduro/nuxt-bitbybit

**Reply by matas-bitbybit-dev** [reply]
@cesarecaoduro I could not really test your app as it was crashing on build time.

I did however set up an example of nuxt 3 here, which currently only uses OCCT, without JSCAD support that will come later I hope
https://github.com/bitbybit-dev/app-examples/tree/main/nuxt/basic

That did cost me quite some time and effort, I really hope it will help you proceed.

**Reply by matas-bitbybit-dev** [reply]
By the way, in [v0.16.3](https://github.com/bitbybit-dev/bitbybit-occt/releases/tag/v0.16.3) there are 3 new methods to get points from wires, namely **pointsOnWireAtLengths**, **pointsOnWireAtEqualLength** and **pointsOnWireAtPatternOfLengths**, maybe some of these will be useful to you. Components are also available in the editors as of today on [bitbybit.dev](https://bitbybit.dev)

**Reply by cesarecaoduro** [reply]
Thanks @matas-bitbybit-dev
I think I did figure what the problem is with the wire I am using (I will try with your implementation)
One of the wires I am trying to evaluate is composed by a series of small segments (some of them less than a unit). I believe that this ends up in being too much computational heavy when I try to evaluate the wire every unit.
![image](https://github.com/user-attachments/assets/a4465802-a12f-4c06-91fb-fb82057c76fa)

In fact the firs 100/200 points are calculate quite quickly. The problem is when it hits the dense area in the image (which is basically made out of several points).

Regardless, thanks for the implementation you have provided!
